compiler/noirc_frontend/src/hir_def/types/arithmetic.rs

```diff
@@ -269,17 +323,16 @@ impl Type {
     fn parse_partial_constant_expr(
         lhs: &Type,
         rhs: &Type,
-    ) -> Option<(Box<Type>, BinaryTypeOperator, FieldElement, FieldElement)> {
+    ) -> Option<(Type, BinaryTypeOperator, FieldElement, FieldElement)> {
         let kind = lhs.infix_kind(rhs);
         let dummy_location = Location::dummy();
         let rhs = rhs.evaluate_to_field_element(&kind, dummy_location).ok()?;
 
-        let Type::InfixExpr(l_type, l_op, l_rhs, _) = lhs.follow_bindings() else {
-            return None;
-        };
+        let lhs = lhs.follow_bindings();
+        let (l_type, l_op, l_rhs, _) = lhs.as_infix_expr()?;
```

Comment:
Is it possible that recursing to find an `InfixExpr` here and/or in `try_unify_by_moving_constant_terms` ends up removing `CheckedCast`'s?

Reply:
Yes - although without doing so we were failing in the `arithmetic_generics` test from failing to catch this case for results which were now cached to include CheckedCast where they previously dropped it